docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,237 @@
+applictions may interact with each other. For example, if Upstream is
+running and I create a project via the CLI, then a good user
+experience is that the Upstream is updated with the new project
+(without refreshing). In RFC-682, it is alluded that a PubSub
+component could solve this, however, that requires the daemon to be
+online. The author prefers not to specify the PubSub component
+here. To bridge the gap, we recommend that a subcommand can add a
+parameter `daemon-mode := offline (default) | online`. If set to
+`online` the command will see if the daemon is online, and if not ask
+it to come online before applying the command.
+
+== Key Access
+
+RFC-682#Signatures outlined the signing of Radicle data. This is essential to
+this RFC as commands that write to the storage generally need access
+to the secret key, whether for signing data or proof of ownership to
+the storage. For most (if not all) read-only commands, no access to
+the secret key is needed.
+
+To achieve the goals of having a read-only storage and using
+`<<ssh-agent>>` for signing, we propose the following should be
+implemented:
+
+. `ssh-agent` compatability with `<<radicle-keystore>>`, which has been
+started in `<<rk-17, radicle-keystore#17>>`.
+. A subcommand `rad-key-ssh-add` for adding one's secret key to the
+`ssh-agent` using the supplied constraints, e.g. ask for password.
+. `radicle-link#461` to have a read-only storage.
+
+== Library/Command Organisation
+
+As proposed here, `radicle-link` will provide a core set of
+functionality. Something this document wishes to standardise is the
+organisation of these components. The author sees two options:
+. One binary that defines the core subcommands
+. Many binary that take advantage of the subcommand mechanism
+
+Pros of *1.*:
+
+* One-stop shop for all core functionality
+* One binary is packaged and distributed
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binary must be packaged and distributed.
+
+The author recommends *1.*. As mentioned, we have a hard requirement
+of not allowing the core commands to be overrided. In *2.* this is
+_always_ possible.
+
+=== `radicle-bins`
+
+Per RFC-682#CLI, all subcommands will be provided as
+libraries. However, this leaves the question of where the binary might
+live. So far, it has been an unspoken formalism that all binaries live
+in `radicle-bins` where we can provide a `Cargo.lock` file. All
+functionality will live in `radicle-link`, including the `main`
+function, and the entry in `radicle-bins` will be a thin wrapper
+around this function.
```

Comment:
It would be nice to scrutinise this more and challenge if this played out well for the seed.